### ISSUE TITLE
Add dtype choice in step type/functions

### ISSUE THREAD

**Issue by thomashirtz** [instadeepai/jumanji#256]
### Is your feature request related to a problem? Please describe
In a personal project, I had to be very efficient in the memory management, my reward were taking a lot of space. In my case I needed to change the reward and the discount to float16 instead of float32. I had to copy over the types file to do the modification locally. However I feel like some use case may need this flexibility.

### Describe the solution you'd like
Give an extra parameter to the step functions to give dtype. (Example with one of them)

```python
def truncation(
    reward: Array,
    observation: Observation,
    discount: Optional[Array] = None,
    extras: Optional[Dict] = None,
    shape: Union[int, Sequence[int]] = (),
    dtype: jnp.dtype = jnp.float32,
) -> TimeStep:
    """Returns a `TimeStep` with `step_type` set to `StepType.LAST`.

    Args:
        reward: array.
        observation: array or tree of arrays.
        discount: array.
        extras: environment metric(s) or information returned by the environment but
            not observed by the agent (hence not in the observation). For example, it
            could be whether an invalid action was taken. In most environments, extras
            is None.
        shape: optional parameter to specify the shape of the rewards and discounts.
            Allows multi-agent environment compatibility. Defaults to () for
            scalar reward and discount.
    Returns:
        TimeStep identified as the truncation of an episode.
    """
    discount = discount if discount is not None else jnp.ones(shape, dtype=dtype)
    extras = extras or {}
    return TimeStep(
        step_type=StepType.LAST,
        reward=reward,
        discount=discount,
        observation=observation,
        extras=extras,
    )
```

I would be happy to do the PR.

---
### Misc
- [x] Check for duplicate requests.

**Comment by sash-a** [comment]
I think that would be a nice addition, happy to review it :smile:

**Comment by sash-a** [comment]
To be honest I think the discounts should probably be booleans while they are stored in the timestep because for me they just indicated end of episode, but I think this would add nice flexibility

**Comment by sash-a** [comment]
My only issue is that this strays from the original [dm_env api](https://github.com/google-deepmind/dm_env/blob/master/dm_env/_environment.py#L148-L157) where it is a float so it can represent both RL discount (gamma) and done. 

Let's definitely add it as an argument, but for the default I'm not sure if boolean or float32 is best @clement-bonnet any thoughts on this?

**Comment by clement-bonnet** [comment]
To my knowledge, having the discount as a float is more common than as a boolean for the reasons you mentioned @sash-a. I would keep it a float unless there are strong reasons to do otherwise :)

**Comment by sash-a** [comment]
Great then if you could add the argument with a default of float32, I'm happy to accept the PR

**Comment by thomashirtz** [comment]
> Great then if you could add the argument with a default of float32, I'm happy to accept the PR

The PR is available for review :)
